scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -2,191 +2,193 @@ package io.iohk.scalanet.peergroup.kademlia
+    nodeRecordIndex.updateAndGet(_.addNodeRecord(nodeRecord))
+    ()
   }
 
-  network.kRequests.foreach {
-    case (FindNodes(uuid, nodeRecord, targetNodeId), responseHandler) =>
-      debug(
-        s"Received request FindNodes(${nodeRecord.id.toHex}, $nodeRecord, ${targetNodeId.toHex})"
-      )
-      add(nodeRecord)
+  private def removeNodeRecord(nodeRecord: NodeRecord[A]): Unit = {
+    nodeRecordIndex.updateAndGet(_.removeNodeRecord(nodeRecord))
+    ()
+  }
 
-      val result = Nodes(uuid, config.nodeRecord, embellish(kBuckets.closestNodes(targetNodeId, config.k)))
+  private def touchNodeRecord(nodeRecord: NodeRecord[A]): Unit = {
+    nodeRecordIndex.updateAndGet(_.touchNodeRecord(nodeRecord))
+    ()
+  }
 
-      sendResponse(result, responseHandler, nodeRecord.routingAddress)
-    case (Ping(uuid, nodeRecord), responseHandler) =>
-      debug(
-        s"Received request Ping(${nodeRecord.id.toHex}, $nodeRecord)"
-      )
-      add(nodeRecord)
-      sendResponse(Pong(uuid, config.nodeRecord), responseHandler, nodeRecord.routingAddress)
+  private def replaceNodeRecord(oldNode: NodeRecord[A], newNode: NodeRecord[A]): Unit = {
+    nodeRecordIndex.updateAndGet(_.replaceNodeRecord(oldNode, newNode))
+    ()
   }
 
-  def get(key: BitVector): Future[NodeRecord[A]] = {
-    debug(s"get(${key.toHex})")
-    getLocally(key).recoverWith { case _ => getRemotely(key) }.recoverWith { case t => giveUp(key, t) }
+  // TODO: parallelism should be configured by library user
+  private val responseTaskConsumer =
+    Consumer.foreachParallelTask[(KRequest[A], Option[KResponse[A]] => Task[Unit])](parallelism = 4) {
+      case (FindNodes(uuid, nodeRecord, targetNodeId), responseHandler) =>
+        debug(
+          s"Received request FindNodes(${nodeRecord.id.toHex}, $nodeRecord, ${targetNodeId.toHex})"
+        )
+        val response = Nodes(uuid, config.nodeRecord, embellish(kBuckets.closestNodes(targetNodeId, config.k)))
+        for {
+          _ <- add(nodeRecord).startAndForget
+          responseTask <- responseHandler(Some(response))
+        } yield responseTask
+
+      case (Ping(uuid, nodeRecord), responseHandler) =>
+        debug(
+          s"Received request Ping(${nodeRecord.id.toHex}, $nodeRecord)"
+        )
+        val response = Pong(uuid, config.nodeRecord)
+        for {
+          _ <- add(nodeRecord).startAndForget
+          responseTask <- responseHandler(Some(response))
+        } yield responseTask
+    }
+
+  /**
+    * Starts handling incoming requests. Infinite task.
+    *
+    * @return
+    */
+  private def startServerHandling(): Task[Unit] = {
+    // asyncBoundary means that we are giving up on observable back pressure and the consumer will need consume
+    // events as soon as available, it is behaviour similar to ConcurrentSubject
+    //TODO: it should be configurable by library user what to do when server does not keep up with requests
```

Comment:
I think the strategy should be `DropNew(k * <some number>)`.

The previous PR was about bounded resource usage for the kbuckets, in terms of k, the 'replication parameter'. There is configuration for that.

It's worth remembering that, as conceived, kademlia uses UDP, so it is already unreliable (and statistical) in nature. We should avoid configuration which leads the user to think that the server is somehow reliable (e.g. bigger queue size) when it is not.

Reply:
after some thought i agree, that default should be `DropNew` (we do not want library causing ooms) , but i think it will be better to expose it as separate config options instead of having `k * someArbitraryNumber`